File: forge/src/main/java/online/ebatel/forge/TpsTracker.java
package online.ebatel.forge;

import net.minecraftforge.event.TickEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;

import java.util.concurrent.ConcurrentLinkedDeque;

public class TpsTracker {
    private static final int SAMPLE_SIZE_1M = 20 * 60;       // 1 minute
    private static final int SAMPLE_SIZE_5M = 20 * 60 * 5;   // 5 minutes
    private static final int SAMPLE_SIZE_15M = 20 * 60 * 15; // 15 minutes

    private static final ConcurrentLinkedDeque<Long> tickTimes = new ConcurrentLinkedDeque<>();
    private static long lastTickTime = System.nanoTime();

    @SubscribeEvent
    public static void onServerTick(TickEvent.ServerTickEvent event) {
        if (event.phase != TickEvent.Phase.END) return;

        long currentTime = System.nanoTime();
        long tickDuration = currentTime - lastTickTime;
        lastTickTime = currentTime;

        tickTimes.addLast(tickDuration);

        while (tickTimes.size() > SAMPLE_SIZE_15M) {
            tickTimes.pollFirst();
        }
    }

    public static double getTps1m() {
        return calculateTps(SAMPLE_SIZE_1M);
    }

    public static double getTps5m() {
        return calculateTps(SAMPLE_SIZE_5M);
    }

    public static double getTps15m() {
        return calculateTps(SAMPLE_SIZE_15M);
    }

    private static double calculateTps(int sampleSize) {
        Long[] samples = tickTimes.toArray(new Long[0]);
        if (samples.length == 0) return 20.0;

        int actualSampleSize = Math.min(sampleSize, samples.length);
        int startIndex = samples.length - actualSampleSize;

        double sum = 0;
        for (int i = startIndex; i < samples.length; i++) {
            sum += samples[i];
        }

        double avgTickTimeNanos = sum / actualSampleSize;
        double avgTickTimeSeconds = avgTickTimeNanos / 1_000_000_000.0;

        double tps = avgTickTimeSeconds > 0 ? 1.0 / avgTickTimeSeconds : 20.0;
        return Math.min(20.0, tps);
    }
}

File: fabric/shared/src/main/kotlin/online/ebatel/fabric/TpsTracker.kt
package online.ebatel.fabric

import java.util.concurrent.ConcurrentLinkedDeque

/**
 * Rolling TPS estimate computed from real tick durations (System.nanoTime deltas at END of each
 * server tick). Same algorithm as the Forge TpsTracker, but driven by Fabric's ServerTickEvents
 * so it stays free of any version-specific MinecraftServer tick-time API.
 */
object TpsTracker {
    private const val SAMPLE_SIZE_1M = 20 * 60        // 1 minute
    private const val SAMPLE_SIZE_5M = 20 * 60 * 5    // 5 minutes
    private const val SAMPLE_SIZE_15M = 20 * 60 * 15  // 15 minutes

    private val tickTimes = ConcurrentLinkedDeque<Long>()

    @Volatile
    private var lastTickTime = System.nanoTime()

    fun onServerTick() {
        val currentTime = System.nanoTime()
        val tickDuration = currentTime - lastTickTime
        lastTickTime = currentTime

        tickTimes.addLast(tickDuration)
        while (tickTimes.size > SAMPLE_SIZE_15M) {
            tickTimes.pollFirst()
        }
    }

    fun getTps1m(): Double = calculateTps(SAMPLE_SIZE_1M)
    fun getTps5m(): Double = calculateTps(SAMPLE_SIZE_5M)
    fun getTps15m(): Double = calculateTps(SAMPLE_SIZE_15M)

    private fun calculateTps(sampleSize: Int): Double {
        val samples = tickTimes.toTypedArray()
        if (samples.isEmpty()) return 20.0

        val actualSampleSize = minOf(sampleSize, samples.size)
        val startIndex = samples.size - actualSampleSize

        var sum = 0.0
        for (i in startIndex until samples.size) {
            sum += samples[i]
        }

        val avgTickTimeNanos = sum / actualSampleSize
        val avgTickTimeSeconds = avgTickTimeNanos / 1_000_000_000.0

        val tps = if (avgTickTimeSeconds > 0) 1.0 / avgTickTimeSeconds else 20.0
        return minOf(20.0, tps)
    }
}
